### PR TITLE
Ensure adequate containerization of lists

### PR DESCRIPTION
**Pull request by spanziera**

### CODE REVIEW

**Comment by spanziera** [comment]
> @nllong While searching for the elements that need to be containerized in the schema, I found some elements that are defined outside of the main path (at the bottom of the code) and referenced multiple times throughout the schema rather than being defined directly within the main path's code.
> 
> Example:
> 
> > Within the main path (reference):
> > <xs:element ref="auc:TenantID" minOccurs="0" maxOccurs="unbounded"/>
> 
> > At the code's bottom, outside the main path (definition):
> > <xs:element name="TenantID">
> 
> In this case, I believe the best solution is to containerize the element once within the main element definition rather than the references. Moreover, I believe it is best to ensure the references point to the parent (plural noun) element, rather than the child (singular noun).
> 
> Please let me know:
> 
> 1. Whether you agree with my approach.
> 2. Whether I should describe how I will handle these special cases within the proposal's implementation.

NOTE: I have found an existing element that follows my approach:
            <xs:element name="PremisesIdentifiers">

**Comment by edward-ellis** [comment]
Although I am generally in favor of consistency, this change is not absolutely necessary and constitutes a major breaking change as it affects about 50 elements.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+<xs:element name="ThermalZoneID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:FacilityType)/auc:Subsections/auc:Subsection/auc:CeilingID/auc:ThermalZoneID -->
+<xs:element name="SpaceID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:FacilityType)/auc:Subsections/auc:Subsection/auc:CeilingID/auc:SpaceID -->
+<xs:element name="ThermalZoneID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:FacilityType)/auc:Subsections/auc:Subsection/auc:FoundationID/auc:ThermalZoneID -->
+<xs:element name="SpaceID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:FacilityType)/auc:Subsections/auc:Subsection/auc:FoundationID/auc:SpaceID -->
+<xs:element name="DeliveryID" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="HVACScheduleID" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="OccupancyScheduleID" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="ScheduleDetails" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="ContactID" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="AnnualSavingsByFuel" minOccurs="0" maxOccurs="unbounded">
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: /Scenarios/Scenario/ScenarioType/PackageOfMeasures/AnnualSavingsByFuel

Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: /Scenarios/Scenario/ScenarioType/PackageOfMeasures/AnnualSavingsByFuel

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+<xs:element name="UtilityMeterNumber" type="xs:string" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="UtilityID" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="Emissions" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="HeatingSource" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="CoolingSource" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="CondenserPlantID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:HVACSystemType)/auc:HeatingAndCoolingSystems/auc:CoolingSource/auc:CoolingSourceType/auc:DX/auc:CondenserPlantID -->
+<xs:element name="Delivery" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="CondenserPlantID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:CoolingPlantType)/auc:Chiller/auc:CondenserPlantID -->
+<xs:element name="CondenserPlantID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:RefrigerationSystemType)/auc:RefrigerationSystemCategory/auc:CentralRefrigerationSystem/auc:CondenserPlantID -->
+<xs:element name="LinkedSystemID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:PumpSystemType)/auc:LinkedSystemID -->
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: I don't think we change this one. I could be swayed though.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+<xs:element name="LinkedSystemID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:FanSystemType)/auc:LinkedSystemID -->
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: I don't think we change this one. I could be swayed though.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+<xs:element name="LinkedSystemID" minOccurs="0" maxOccurs="unbounded"> <!-- //element(*,auc:MotorSystemType)/auc:LinkedSystemID -->
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: I don't think we change this one. I could be swayed though.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+1. Search for all occurrences of "maxOccurs" where the value is set to a number higher than 1 or to "unbounded."
+2. Of the found elements, discard those whose name is a singular noun and whose parent element is named with the same noun in the plural form.
+3. For all elements left:
+	* Ensure their names are plural nouns.
+	* Create a child element for them that is named with the same noun in singular form.
+
+List 1, 2, ad 3 are the proposed lists of elements to be changed. These elements were identified using steps 1 and 2.  Note that elements with repeated names each include their schema model path beside them as a comment.
+
+
+List 1: elements to be changed that do not represent any special cases.
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: These look good. I made a couple exceptions in the list which need to be reviewed.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+<xs:element name="CondenserPlant" type="auc:CondenserPlantType" minOccurs="0" maxOccurs="unbounded">
+<xs:element name="LinkedScheduleID" minOccurs="0" maxOccurs="unbounded"> <!-- auc:LinkedPremises/auc:Site/auc:LinkedSiteID/auc:LinkedScheduleID -->
+<xs:element name="LinkedScheduleID" minOccurs="0" maxOccurs="unbounded"> <!-- auc:LinkedPremises/auc:Facility/auc:LinkedFacilityID/auc:LinkedScheduleID -->
+<xs:element name="LinkedScheduleID" minOccurs="0" maxOccurs="unbounded"> <!-- auc:LinkedPremises/auc:Subsection/auc:LinkedSubsectionID/auc:LinkedScheduleID -->
+<xs:element name="LinkedScheduleID" minOccurs="0" maxOccurs="unbounded"> <!-- auc:LinkedPremises/auc:ThermalZone/auc:LinkedThermalZoneID/auc:LinkedScheduleID -->
+<xs:element name="LinkedScheduleID" minOccurs="0" maxOccurs="unbounded"> <!-- auc:LinkedPremises/auc:Space/auc:LinkedSpaceID/auc:LinkedScheduleID -->
+```
+
+
+List 2: elements that are defined outside of the main schema path (at the bottom of the code) and are referenced throughout the schema code. For mor information, reference the following pull request comment: https://github.com/BuildingSync/schema/pull/67#issuecomment-456912360.
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: Yes, making the changes to the main part of the schema makes sense.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+List 2: elements that are defined outside of the main schema path (at the bottom of the code) and are referenced throughout the schema code. For mor information, reference the following pull request comment: https://github.com/BuildingSync/schema/pull/67#issuecomment-456912360. 
+
+```xml
+<xs:element name="TenantID">
+<xs:element name="EnergyUseByFuelType">
+```
+
+
+List 3: unique cases to be reviewed by the BuildingSync technical lead to define approach.
```
Comment (nllong) [inline proposals/2019/Ensure adequate containerization of lists.txt]: What about making this look like the following?

image

**Comment by nllong** [comment]
@edward-ellis Thanks for reviewing this PR. The request to make the containers consistent come from PNNL awhile back and we think that now is the best time to do that as we roll out 1.0. Can you chat with @markborkum to make sure that this makes sense? There are a lot of touched elements, but it will definitely make it easier for others to adopt having it be consistent.

**Comment by aqwolfe** [comment]
@nllong starting on this item. I just merged this branch from develop so that it has changes made in #64. It doesn't have changes made in #58 but those were non-breaking changes that shouldn't interfere with these changes to containerization.

File: proposals/2019/Ensure adequate containerization of lists.txt [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+
+*Unique case C: The following element is a child of another element of the same exact name. Since this one is the child element, it should be in singular form. Proposed approach: remove the 's' at the end of the following element's name. 
+
+```xml
+<xs:element name="DuctSystems" type="auc:DuctSystemType" maxOccurs="unbounded"/>
+```
+
+
+*Unique case D: The parents of the following elments are not named with the plural version of thir children's names. However, they are very simmilar. Proposed approach: for each of the listed elements, name the respective parent using the plural version of the listed element's name.
```
[the remark this answers is not in the document]
Comment (aqwolfe) [reply proposals/2019/Ensure adequate containerization of lists.txt]: ok we will leave this as is

File: examples/ASHRAE 211 Export.xml [hunk 9]
```diff
@@ -386,39 +386,49 @@
         <HVACSystems>
           <HVACSystem>
             <Plants>
+            <HeatingPlants>
               <HeatingPlant ID="HeatiingPlant1">
```
Comment (nllong) [inline examples/ASHRAE 211 Export.xml]:
```suggestion
              <HeatingPlant ID="HeatingPlant1">
```

**Comment by nllong** [comment]
> Although I am generally in favor of consistency, this change is not absolutely necessary and constitutes a major breaking change as it affects about 50 elements.

We decided to implement these changes for 1.0. Yes, this touches many elements, but I think that many of the elements may not be used yet by asset score. Having this in version 1.0 will prevent us from having to update to version 2.0 in the very near future.

**Pull request merged**